### PR TITLE
[Distributed][Model] Rank-based Component Creation for Pipeline Parallelism Memory Optimization

### PR DESCRIPTION
**Pull request by wushidonguc**
Prior to this change, the Llama model initialization in Pipeline Parallel scenarios would create all components and layers for every rank, leading to unnecessary memory overhead. This pull request optimizes memory usage by creating certain components only on the relevant ranks.

The primary changes include:
1. Creating the embedding layer on the first rank only.
2. Creating the norm layer on the last rank only.
3. Creating the lm head, logits processor, and sampler components on the last rank only.

By selectively creating components on the relevant ranks, this optimization reduces the overall memory footprint. Testing with `--pipeline-parallel-size 8` showed up to 25% per-rank memory savings for the Llama-3-70B model. This memory optimization can potentially enable higher throughput by allowing more memory to be utilized for serving.

This optimization enables serving larger models and running inference on resource-constrained environments by leveraging Pipeline Parallelism more efficiently.

Testing:
- Benchmarks have been conducted to measure the memory savings and performance impact of this optimization.

Please review the changes and provide feedback or suggestions for further improvements.

### CODE REVIEW

**Comment by sfc-gh-hazhang** [comment]
looks nice. I'll do a test on H100 setups and confirm improvement.

File: vllm/model_executor/models/llama.py [hunk 1]
```diff
@@ -276,7 +277,10 @@ def __init__(
                 nn.Identity()
                 for _ in range(self.end_layer, config.num_hidden_layers)
             ])
-        self.norm = RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        if get_pp_group().is_last_rank:
+            self.norm = RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        else:
+            self.norm = None
```
Comment (youkaichao) [inline vllm/model_executor/models/llama.py:283]: I suppose `norm` should take only a small fraction of GPU memory. is it worth the effort?

File: vllm/model_executor/models/llama.py [hunk 2]
```diff
@@ -369,26 +373,31 @@ def __init__(
...
+                config.hidden_size,
+                org_num_embeddings=config.vocab_size,
+                padding_size=DEFAULT_VOCAB_PADDING_SIZE
+                # We need bigger padding if using lora for kernel
+                # compatibility
+                if not lora_config else lora_config.lora_vocab_padding_size,
+                quant_config=quant_config,
+            )
+            if config.tie_word_embeddings:
+                self.lm_head.weight = self.model.embed_tokens.weight
```
Comment (youkaichao) [inline vllm/model_executor/models/llama.py:391]: can we move the logic into `ParallelLMHead` so that the rest models can benefit from it as well?

**Comment by youkaichao** [comment]
thanks for the contribution! the overall idea looks good to me.

we actually use the minimum of the number of blocks across all processes to determine the final block size, which means it is only effective if we can reduce the lower bound of model memory usage.

In this case, you:
- removed word embedding for the first pp rank
- removed both word embedding and lm head for the rest ranks
- removed lm head for the last rank

Overall, the minimum memory usage of the model is reduced by about the size of a word embedding or an lm head layer (typically the same size).

It would be better if you can move this logic into common layers, so that the rest models can directly benefit from it. I added `PPMissingLayer` in https://github.com/vllm-project/vllm/pull/6406 recently, which might help. My goal is to reduce the code change to support pp, to make sure the code is not too intrusive and people can find it easy to integrate.

One niche case, is you need to take care of models that use tie embeddings.

**Comment by andoorve** [comment]
Thanks for your change! This should bring some improvement. I'll let @sfc-gh-hazhang report back.

**Comment by wushidonguc** [comment]
> thanks for the contribution! the overall idea looks good to me.
> 
> we actually use the minimum of the number of blocks across all processes to determine the final block size, which means it is only effective if we can reduce the lower bound of model memory usage.
> 
> In this case, you:
> 
>     * removed lm head for the first pp rank
> 
>     * removed both word embedding and lm head for the rest ranks
> 
>     * removed word embedding for the last rank
> 
> 
> Overall, the minimum memory usage of the model is reduced by about the size of a word embedding or an lm head layer (typically the same size).
> 
> It would be better if you can move this logic into common layers, so that the rest models can directly benefit from it. I added `PPMissingLayer` in #6406 recently, which might help. My goal is to reduce the code change to support pp, to make sure the code is not too intrusive and people can find it easy to integrate.
> 
> One niche case, is you need to take care of models that use tie embeddings.

Thanks for the feedback and suggestions!

Instead of modifying `VocabParallelEmbedding` and `ParallelLMHead`, I propose keeping the pipeline parallelism logic within the model-specific classes. We can introduce helper methods to encapsulate the rank-aware logic. These helpers would return the appropriate instances based on the current rank.

This way, `VocabParallelEmbedding` and `ParallelLMHead` remain generic and reusable across models. The pipeline parallelism logic stays in the model-specific classes, making it easier to maintain and extend.

Please let me know if this approach works for you.

**Comment by youkaichao** [comment]
> The pipeline parallelism logic stays in the model-specific classes, making it easier to maintain and extend.

this idea looks good to me. I'd like to make the code change as small as possible, and it would be better if we can make the code intuitive.

File: vllm/model_executor/models/llama.py [hunk 3]
```diff
@@ -257,11 +257,12 @@ def __init__(
                       (lora_config.max_loras or 1)) if lora_config else 0
         self.vocab_size = config.vocab_size + lora_vocab
         self.org_vocab_size = config.vocab_size
-        self.embed_tokens = VocabParallelEmbedding(
-            self.vocab_size,
-            config.hidden_size,
-            org_num_embeddings=config.vocab_size,
-        )
```
Comment (youkaichao) [inline vllm/model_executor/models/llama.py:264]: How about adding one line?

`self.embed_tokens = PPMissingLayer()`

The `PPMissingLayer` exists in `vllm.model_executor.models.utils`

File: vllm/model_executor/models/llama.py [hunk 4]
```diff
@@ -257,11 +257,12 @@ def __init__(
...
+        if get_pp_group().is_first_rank:
+            self.embed_tokens = VocabParallelEmbedding(
+                self.vocab_size,
+                config.hidden_size,
+                org_num_embeddings=config.vocab_size,
+            )
```
Comment (youkaichao) [inline vllm/model_executor/models/llama.py:265]: How about adding:

```
else:
    self.embed_tokens = PPMissingLayer()
```

The `PPMissingLayer` exists in `vllm.model_executor.models.utils`

This helps weight loading logic to correctly skip weights.

**Comment by wushidonguc** [comment]
@youkaichao I've rebased the branch on the latest `main` and integrated the `PPMissingLayer` component to handle weight skipping. No helper functions are added at this stage, but they can be introduced in a separate pull request if deemed necessary.

Please review the changes and provide any feedback or concerns you may have. I tested the integration of `PPMissingLayer` locally, and it seems to be working as expected.

File: vllm/model_executor/models/llama.py (hunk 4 above)
Comment (wushidonguc) [reply vllm/model_executor/models/llama.py:265]: Done

File: vllm/model_executor/models/llama.py [hunk 5]
```diff
@@ -360,26 +366,31 @@ def __init__(
...
+            self.lm_head = ParallelLMHead(
+                self.unpadded_vocab_size,
+                config.hidden_size,
+                org_num_embeddings=config.vocab_size,
+                padding_size=DEFAULT_VOCAB_PADDING_SIZE
+                # We need bigger padding if using lora for kernel
+                # compatibility
+                if not lora_config else lora_config.lora_vocab_padding_size,
+                quant_config=quant_config,
+            )
```
Comment (youkaichao) [inline vllm/model_executor/models/llama.py:382]: I think you should only put `if-else` for this layer, `LogitsProcessor` and `Sampler` don't have parameters. Let's make the change as small as possible.

File: vllm/model_executor/models/llama.py [hunk 6]
```diff
@@ -257,17 +257,23 @@ def __init__(
                       (lora_config.max_loras or 1)) if lora_config else 0
         self.vocab_size = config.vocab_size + lora_vocab
         self.org_vocab_size = config.vocab_size
-        self.embed_tokens = VocabParallelEmbedding(
-            self.vocab_size,
-            config.hidden_size,
-            org_num_embeddings=config.vocab_size,
-        )
+        if get_pp_group().is_first_rank:
```
Comment (youkaichao) [inline vllm/model_executor/models/llama.py:260]: `if get_pp_group().is_first_rank or (config.tie_word_embeddings and get_pp_group().is_last_rank)` ? we need to take care of tie word embedding as well.

File: vllm/model_executor/models/llama.py (hunk 5 above)
Comment (wushidonguc) [reply vllm/model_executor/models/llama.py:383]: Excluded `LogitsProcessor` and `Sampler` in else

File: vllm/model_executor/models/llama.py [hunk 7]
```diff
@@ -360,26 +367,29 @@ def __init__(
...
+            if config.tie_word_embeddings:
+                self.lm_head.weight = self.model.embed_tokens.weight
 
-        logit_scale = getattr(config, "logit_scale", 1.0)
-        self.logits_processor = LogitsProcessor(self.unpadded_vocab_size,
+            logit_scale = getattr(config, "logit_scale", 1.0)
+            self.logits_processor = LogitsProcessor(self.unpadded_vocab_size,
                                                 config.vocab_size, logit_scale)
-        self.sampler = Sampler()
+            self.sampler = Sampler()
```
Comment (youkaichao) [inline vllm/model_executor/models/llama.py:390]: some niche details, I think it would be better to keep these two lines untouched.

your code should just add `if-else` around:

```python
        self.lm_head = ParallelLMHead(
            self.unpadded_vocab_size,
            config.hidden_size,
            org_num_embeddings=config.vocab_size,
            padding_size=DEFAULT_VOCAB_PADDING_SIZE
            # We need bigger padding if using lora for kernel
            # compatibility
            if not lora_config else lora_config.lora_vocab_padding_size,
            quant_config=quant_config,
        )
        if config.tie_word_embeddings:
            self.lm_head.weight = self.model.embed_tokens.weight
```

Comment (wushidonguc) [reply vllm/model_executor/models/llama.py:390]: I understand your perspective about keeping those lines untouched. However, placing `LogitsProcessor` and `Sampler` within the if-else statement avoids unnecessary object creation for ranks other than the last rank.

**Review by youkaichao — APPROVED** [review]
test failure in https://buildkite.com/vllm/ci-aws/builds/5017#0190be33-93a3-48a3-a443-90c80d0b63df can be ignored, I just disabled them because they are flaky. need to wait for @andoorve to fix.

as long as other tests are passed, we can merge this PR.

thanks for the contribution!
